### ISSUE TITLE
Buffer Overflow?

### ISSUE THREAD

**Issue by wilwade** [privatenumber/tsx#652]
### Acknowledgements

- [X] I read the documentation and searched existing issues to avoid duplicates
- [X] I understand this is a **bug tracker** and anything other than a proven bug will be closed
- [X] I understand this is a free project and relies on community contributions
- [X] I read and understood the [Contribution guide](https://github.com/privatenumber/tsx/blob/master/CONTRIBUTING.md)

### Minimal reproduction URL

https://gist.github.com/wilwade/66bc4614fdaa3a349c5f25b7164d12bc

### Problem & expected behavior (under 200 words)

tsx Version: v4.19.1, v4.10.2
Note: This ALSO happens with ts-node v10.9.2

tldr: When using a buffer, the `buffer.buffer` is getting filled up with stuff it should not be.

```
let decoder = new TextDecoder("utf-8");
const zerosArray = new Array(1024).fill(0);
// @ts-ignore
const buf = Buffer.from(zerosArray);
```

### Bugs are expected to be fixed by those affected by it

- [ ] I'm interested in working on this issue

### Compensating engineering work will speed up resolution and support the project

- [ ] I'm willing to offer $10 for financial support

**Comment by shannonwells** [comment]
More info:
* the behavior also occurs for `Buffer.copyBytesFrom`
* the Buffer's `ArrayBuffer` is always 8192 in length, with a byteLength & length of whatever the original array length was, and a byteOffset = 4568 or something like it.
* Instance method `toString()` outputs the correct values anyway
* Instance method `entries()` also outputs the correct values
* Does not occur when the same code is compiled into JS and then executed.

The workaround is to use Buffer.alloc with a TypedArray.  Example:
```
const uint8array = new Uint8Array([0,1,2,3,5,8,13]);
const buf = Buffer.alloc(uint8array.length, uint8array)
```

The above results in the expected values in the Buffer's ArrayBuffer, with a length and byteLength of 7, byteOffset = 0 as expected.

**Comment by diracdeltas** [comment]
Any update on this?

**Comment by fallaciousreasoning** [comment]
Okay, this seems to be related to `Buffer.poolSize` - the only reason this comes up in `ts-node` and `tsx` and not in the following JS:
```js
const buf1 = Buffer.from('hello')
console.log(new TextDecoder().decode(buf1.buffer))
// outputs "hello"
```

is because there's no other data in a buffer in that JS - its pretty straightforward to duplicate with the following:

```js
const buf0 = Buffer.from("I'm some other text that you didn't expect!")
const buf1 = Buffer.from('hello')
console.log(new TextDecoder().decode(buf1.buffer))
// outputs "/I'm some other text that you didn't expect!hello"
```

You can prevent this from happening with (what I'm sure is a pretty significant performance penalty) but setting `Buffer.poolSize = 1` before creating your buffer:

```js
const buf0 = Buffer.from("I'm some other text that you didn't expect!")
Buffer.poolSize = 1
const buf1 = Buffer.from('hello')
console.log(new TextDecoder().decode(buf1.buffer))
// outputs "hello"
```

If anything, I'd say this is a bug in `TextDecoder` from `node:utils` but it maybe just be working as intended.

**Comment by privatenumber** [comment]
Thanks for the report. I reproduced the exact gist under both current plain Node and tsx.

Both expose an 8 KiB pooled backing `ArrayBuffer`. tsx makes the unrelated bytes more visible because Node loads additional modules before user code. A fresh Node process may appear clean only by coincidence; one prior small Buffer allocation reproduces the visible behavior without tsx.

This is expected Node `Buffer` behavior, not an overflow or a tsx/ts-node bug. Small `Buffer.from(...)` results are views into Node's shared allocation pool. `buf.buffer` exposes the complete backing store; only the range from `buf.byteOffset` through `buf.byteOffset + buf.byteLength` belongs to `buf`.

The simplest fix is to pass the Buffer directly to `TextDecoder`:

```js
decoder.decode(buf)
```

If an exact `ArrayBuffer` is required:

```js
const arrayBuffer = buf.buffer.slice(
  buf.byteOffset,
  buf.byteOffset + buf.byteLength,
)
```

Node documents this behavior under [`buf.byteOffset`](https://nodejs.org/api/buffer.html#bufbyteoffset).

<details>
<summary>How this was investigated</summary>

The behavior can be inspected directly:

```js
const prior = Buffer.from('unrelated pooled bytes')
const buf = Buffer.from(new Array(1024).fill(0))

console.log({
  length: buf.length,
  byteOffset: buf.byteOffset,
  backingLength: buf.buffer.byteLength,
  sharesBackingStore: prior.buffer === buf.buffer,
})
```

Across Node 18.20.8, 20.20.2, 22.22.3, 24.15.0, and 26.1.0:

| Runtime | Logical length | Backing length | Example offset |
| --- | ---: | ---: | ---: |
| Plain Node 18 | 1,024 | 8,192 | 16 |
| tsx on Node 18 | 1,024 | 8,192 | 3,760 |
| Plain Node 24 | 1,024 | 8,192 | 1,632 |
| tsx on Node 24 | 1,024 | 8,192 | 6,480 |

Every logical byte was correct. Only bytes outside the Buffer view were unrelated.

A size sweep found the exact pooling boundary:

| Allocation size | Backing length | Pooled |
| ---: | ---: | --- |
| 4,095 | 8,192 | Yes |
| 4,096 | 4,096 | No |
| 4,097 | 4,097 | No |

To identify why tsx changes the offset, I preloaded instrumentation around `Buffer.from()`, `Buffer.allocUnsafe()`, and `Buffer.concat()`:

```sh
NODE_OPTIONS=--require=./trace-buffer.cjs tsx buffer-test.ts
```

The call stacks showed Node's ESM loader performing nine pooled `fs.readFileSync()` allocations for tsx runtime modules. Their aligned slices occupied offsets 0 through 6,479, so the user's Buffer began at 6,480. Running through `node --import tsx` produced the same result, ruling out the tsx CLI and IPC layer.

I also captured Node CPU and heap profiles with `--cpu-prof` and `--heap-prof`. Buffer conversion was not a CPU hotspot; the difference was simply which normal startup allocations had already used the slab.

Seven-process allocation benchmarks showed why tsx should not disable pooling globally:

| Operation | Median throughput |
| --- | ---: |
| `Buffer.from(array)`, normal pooling | 2.09M/s |
| Same operation with pooling disabled | 1.01M/s |

For the reported decode operation:

| Input | Median throughput |
| --- | ---: |
| `decoder.decode(buf)` | 9.32M/s |
| Decode a copied exact `ArrayBuffer` | 2.31M/s |

Passing the Buffer directly is therefore both correct and approximately four times faster than copying its backing range before every decode.

</details>

Changing tsx would require overriding Node's global Buffer allocator, affecting every dependency and reducing allocation performance. I'm closing this as working as intended.
